### ISSUE TITLE
Augment logger translations with metadata

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#7494]
The following metadata should just work:

  * [x] `:name` ([source](https://github.com/appsignal/appsignal-elixir/blob/e3d304e2424a8e648f66ee1f23cc7da427bacea2/lib/appsignal/error_handler.ex#L81-L84)). This option appears optionally in `:registered_name` in proc_lib crash reports and it may be present in behaviours as an atom or PID

The following metadata requires new formatter rules:

  * [x] `:initial_call` ([source](https://github.com/getsentry/sentry-elixir/blob/747964b81dab4a3f79a154d3f918ba3c9fd95c2a/lib/sentry/logger.ex#L121)) - available only in proc_lib crash reports.

The following metadata is always ignored unless explicitly fetched by the logger backend:

  * [x] `:crash_reason` as `{reason, stack}` - ([source](https://github.com/getsentry/sentry-elixir/blob/747964b81dab4a3f79a154d3f918ba3c9fd95c2a/lib/sentry/logger.ex#L108-L115)) ([source](https://github.com/elixir-addicts/rollbax/blob/7ff1181b0f8e2ed1878b5bb4c8496f9eda144510/lib/rollbax/reporter/standard.ex#L121-L140)) ([source](https://github.com/jarednorman/bugsnag-elixir/blob/master/lib/bugsnag/logger.ex#L21-L23)). It is available as `{kind, reason, stack}` in `proc_lib` and it is in the return of `format_reason` in behaviours

**Comment by josevalim** [comment]
We can also use the first stacktrace entry to fill in the module and function.

**Comment by mitchellhenke** [comment]
happy to take this if nobody is actively working on it. would be able to get started on it this week 🙂

**Comment by josevalim** [comment]
@mitchellhenke I am already working on it :) but reviews will be really appreciated when I send a PR. :)

**Comment by mitchellhenke** [comment]
would be glad to!

**Comment by mitchellhenke** [comment]
comment updated

**Comment by josevalim** [comment]
We want to include some of it (pid, registered_name, error_info, etc) but not all of it. It should be easy to add new information at any time though. Is there anything in particular you would like to add?

**Comment by mitchellhenke** [comment]
dictionary is particularly useful I think. It would include `logger_metadata` which could even be pulled out and passed to the eventual Logger call as well?

**Comment by josevalim** [comment]
@mitchellhenke oh, genius. We should extract the logger_metadata from the dictionary. I wouldn't include other dictionary keys though. Is that ok?

**Comment by mitchellhenke** [comment]
@josevalim It's a grey area to me. My argument in favor would be that it's information included in a log event from Erlang, so it could be passed along to all of the Logger backends to do with what they like.

I also understand not wanting to pass around the whole thing and having to deal with issues that could arise from that.

If other process dictionary keys weren't included, my use case would have to come up with a process-based dictionary somewhere or make use of`Logger.metadata` to store/access the same information.

**Comment by josevalim** [comment]
The problem is that anything can store something in the dictionary, such as the rand module. I think we can be conservative here. In the worst scenario, you can provide your own translation module and extract metadata before Logger puts its hands on it. :)

**Comment by mitchellhenke** [comment]
I hadn't thought of a custom translator, but that would work as well. My preference would be to allow Logger backends to receive all of that information, but I understand staying conservative in handling that information.

This discussion has been really helpful, and I appreciate you taking the time to explain things and answer my questions. Looking forward to the PR! 🙂

**Comment by josevalim** [comment]
Apologies but I had to focus on other stuff and I was not able to progress on this.

@mitchellhenke I do have one question though. I see that you also have a plug: https://github.com/getsentry/sentry-elixir/blob/747964b81dab4a3f79a154d3f918ba3c9fd95c2a/lib/sentry/plug.ex#L1

How do you avoid reporting the errors twice? One from plug and another from the process crash?

**Comment by mitchellhenke** [comment]
@josevalim No worries about having other stuff to focus on, completely understand!

This is something that's currently not ideal, but Sentry's current `error_logger` handler doesn't handle the message sent on a Plug request process crash.

I've considered expanding what events the handler generates reports from to minimize other configuration (like the Plug) that is necessary, but stacktrace information is generally taken from the arguments to the formatted message at the moment. My understanding is that if the format of the messages were to change it may cause issues with fetching that information?

For example, I see that when a Plug request crashes, `error_logger` passes along the message `Ranch listener ~p had connection process started with ~p:start_link/4 at ~p exit with reason: ~999999p~n'`. It's a list of 4 pieces of data, the final one which contains the structured stacktrace. That last one is a two-element tuple in which I could match on `{{error, stack}, _}` and pull out the stacktrace.

I wasn't sure how often the format may change, so I've avoided depending on it in favor of solutions like the Plug.

If your time is limited because of other things, I'd be glad to take this on and take it in whichever direction you think is best in solving the original issue 🙂

**Comment by josevalim** [comment]
@mitchellhenke the PR #7649 adds support for returning metadata in the translator. We should merge it tomorrow. If you would like to go ahead and return the relevant metadata after it is merged a PR will be very welcome!

Sorry for the slow replies but some of the decision around this area was pending on OTP 21 but we should be good to go now.

**Comment by josevalim** [comment]
@mitchellhenke also please note that Elixir v1.7 and Erlang/OTP 21 will automatically read the Logger metadata on log messages. So this particular issue has already been solved. We can now focus on extracting the metadata from the error reports themselves.

**Comment by mitchellhenke** [comment]
@josevalim awesome, thank you!

I do have a couple questions if you don't mind. What do you mean by Logger metadata will automatically be read? Is there somewhere to read what is all changing for logging in OTP 21?

Thanks again 🙂

**Comment by mitchellhenke** [comment]
@josevalim thanks for the link! I am hoping to start taking a look into it this week.

**Comment by josevalim** [comment]
@mitchellhenke we could allow `{:none, metadata}` but in this case you could format the message as well by calling `:io_lib.format(msg, args)`.

**Comment by mitchellhenke** [comment]
Ah ok, I was hesitant to do any message formatting as I wasn't sure if the translator should be strictly concerned with metadata.

Do you think that the structure in the example is workable for including metadata in Logger events from `error_logger`, and is it something that you think would be included in core?

**Comment by josevalim** [comment]
@mitchellhenke I believe Elixir should provide some metadata by default, at least the ones defined in the issue topic.

**Comment by josevalim** [comment]
Hi @mitchellhenke! This week we are closing in on the issues tracker to do a release candidate for v1.7.0. Did you move forward on this front? If you haven't finished yet but already has something working, please let us know.

**Comment by mitchellhenke** [comment]
@josevalim yeah! I think I've got most of the structure in place, but ran out of time this week to add it across all of the translator tests.  I'll be able to finish it up tomorrow if that's alright?

Current progress can be seen here (happy to receive feedback): https://github.com/elixir-lang/elixir/compare/master...mitchellhenke:logger-metadata?expand=1

**Comment by josevalim** [comment]
@mitchellhenke looks excellent to me! :D

I think we could change the translator tests in their entirety to use the new backend but we should do it later. :)

**Comment by josevalim** [comment]
I have played a bit with `initial_call` and unfortunately it is very inconsistent. For example, for supervisors, it returns `{supervisor, Module, 1}`. My suggestion is that, instead of including `initial_call`, we should include `initial_module` or `proc_module`, that only includes the module name and does not need extra rules.

**Comment by josevalim** [comment]
@michalmuskala good point. I think for now we can go ahead and change the supervisor initial_call to a proper MFA and one day we may no longer need to do it.

**Comment by josevalim** [comment]
@mitchellhenke since we are closing on a release candidate for v1.7, i went ahead and implemented the remaining initial_call and registered_name metadata. Thanks for the great test logger and the crash report implementation.  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

Closing in favor of #7844.

**Comment by mitchellhenke** [comment]
@josevalim sounds good! Happy to have been able to contribute a little bit 🙂 

I'll definitely play around with the new translators/metadata when the release candidate comes out.
